src/Chunk.ts

```diff
@@ -592,10 +586,26 @@ export default class Chunk {
 			}
 
 			let reexports = reexportDeclarations[dep.id];
+			const isExternal = !!(<ExternalModule>dep).isExternal;
+			let exportsNames: boolean, exportsNamespace: boolean, exportsDefault: boolean;
+			if (isExternal) {
+				exportsNames = (<ExternalModule>dep).exportsNames;
+				exportsNamespace = (<ExternalModule>dep).exportsNamespace;
+				exportsDefault = 'default' in (<ExternalModule>dep).declarations;
+			} else {
+				exportsNames = true;
+				// we don't want any interop patterns to trigger
+				exportsNamespace = false;
+				exportsDefault = false;
+			}
+
 			dependencies.push({
 				id: dep.id,
 				name: dep.name,
 				isChunk: !(<ExternalModule>dep).isExternal,
+				exportsNames,
+				exportsNamespace,
+				exportsDefault,
 				reexports,
 				imports
 			});
```

Comment:
Unfortunately the scope of all variables in this function follows the whole function. See for example `reexportDeclarations` is referenced by the dependencies loop, and `imports` being referenced from above as well.

Reply:
No, this is exactly the kind of confusion I would like to avoid with this extraction. I was talking about the forEach loop starting at L.573. Except for the variables `reexportDeclarations` as an input and `dependencies` as an output, this loop is completely self-contained (as I tried out in my IDE). This is easily overlooked as long as it is just somewhere inside the larger function scope.
The same goes for the for-loop starting at L.553 which only produces `reexportDeclarations`.

So to repeat, my suggestion is:
* extract L.549-L.569 into a private method `getReexportDeclarations()`.
* extract L.571-L.610 into a private method `getModuleDeclarationDependencies()` which itself calls `getReexportDeclarations()` to get the re-exports.

So everything that remains in this function is the generation of `exports` which may or may not be extracted itself for symmetry reasons.

Reply:
Nice!